File: projects/rxtwo/src/main/java/com/azunitech/rx/rxtwo/persistent/ProductGatewayImpl.java
package com.azunitech.rx.rxtwo.persistent;

import com.azunitech.rx.rxtwo.domain.Product;
import com.azunitech.rx.rxtwo.domain.ProductGateway;
import com.azunitech.rx.rxtwo.persistent.model.ProductDocument;
import com.azunitech.rx.rxtwo.persistent.repository.UserRepository;
import com.azunitech.rx.rxtwo.persistent.utils.Utils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.stream.Collectors;

@Component
class ProductGatewayImpl implements ProductGateway {
  UserRepository repository;

  @Autowired
  public ProductGatewayImpl(UserRepository repository) {
    this.repository = repository;
  }

  @Override
  public List<Product> findProduct(FindProductQuery findUserQuery) {
    return repository.findByName(findUserQuery.getName()).stream()
        .map(ProductDocument::toDomain)
        .collect(Collectors.toList());
  }

  @Override
  public Product createProduct(Product product) {
    ProductDocument document = Utils.converter.apply(product);
    return repository.insert(document).toDomain();
  }
}

File: projects/cleancode/src/main/kotlin/com/azunitech/cleancode/persistent/ProductGatewayImpl.kt
package com.azunitech.cleancode.persistent

import com.azunitech.cleancode.domain.FindProductsQuery
import com.azunitech.cleancode.domain.Product
import com.azunitech.cleancode.domain.ProductGateway
import org.springframework.stereotype.Component
import java.util.stream.Collectors

@Component
class ProductGatewayImpl(private val productRepository: ProductRepository? = null) : ProductGateway {
    override fun create(product: Product): String {
        val saved = productRepository!!.save(product.toNewDocument())
        return saved.businessKey
    }

    override fun findProducts(query: FindProductsQuery): List<Product> {
        return productRepository!!.findByNameContains(query.nameContains).stream().map { it.toDomain() }
                .collect(Collectors.toList())
    }
}

fun Product.toNewDocument(): ProductDocument {
    return ProductDocument(null, id, name);
}